src/native/libs/System.Native/pal_uid.c

```diff
@@ -237,3 +237,54 @@ int32_t SystemNative_GetGroups(int32_t ngroups, uint32_t* groups)
 
     return getgroups(ngroups, groups);
 }
+
+int32_t SystemNative_GetUName(uint32_t uid, char* buffer, int32_t bufferSize)
+{
+    assert(buffer != NULL || bufferSize == 0);
+    assert(bufferSize >= 0);
+
+    if (bufferSize <= 0)
+    {
+        errno = EINVAL;
+        return -1;
+    }
+
+    errno = 0;
+    struct passwd *pwd;
+    while ((pwd = getpwuid(uid)) == NULL && errno == EINTR);
+
+    size_t unameLength = strlen(pwd->pw_name);
+    if (errno < 0 || unameLength <= 0 || unameLength > (size_t)bufferSize)
+    {
+        buffer = NULL;
+        return -1;
+    }
+    SafeStringCopy(buffer, Int32ToSizeT(bufferSize), pwd->pw_name);
+    return (int32_t)unameLength;
+}
+
+int32_t SystemNative_GetGName(uint32_t gid, char* buffer, int32_t bufferSize)
+{
+    assert(buffer != NULL || bufferSize == 0);
+    assert(bufferSize >= 0);
+
+    if (bufferSize <= 0)
+    {
+        errno = EINVAL;
+        return -1;
+    }
+
+    errno = 0;
+    struct group *grp;
+    while ((grp = getgrgid(gid)) == NULL && errno == EINTR);
```

Comment:
getgrgid, unlike getgrgid_r, is [not necessarily thread safe](https://www.man7.org/linux/man-pages/man3/getgrgid.3p.html?msclkid=5941b436d0b211ec95adc3995ff60954). I guess it uses some static scratch pad or something like that.

If I understand correctly then, this either needs a lock, or use getgrid_r instead -- which might have to be protected with a  HAVE_GETGRGID_R check in the makefiles. 

Similarly for getpwuid and getpwuid_r.

@tmds do I understand correctly? Also, if we use the _r functions do we need to test for their existence, or can we assume it?

Reply:
eg, https://github.com/mono/mono/blob/main/mono/metadata/mono-security.c#L112

Reply:
> `The getgrgid() function need not be thread-safe.`

This sounds to me like they are saying "you don't need to worry about thread safety with this method". But maybe my interpretation is wrong. What do you read from that?

Reply:
It's a specification for getgrgid(). If you implement it, you need not make it thread safe. Put another way, if you call it, you can't rely on it being threadsafe. It seems it is better to use _r. But @tmds can advise.

Reply:
> It seems it is better to use _r

Yes, for thread-safety use the `*_r` functions. The other ones may rely on static data that is shared across threads.

Reply:
> The other ones may rely on static data that is shared across threads.

And locking is an insufficient mitigation because we don't control all possible call sites.

Reply:
Fixed.